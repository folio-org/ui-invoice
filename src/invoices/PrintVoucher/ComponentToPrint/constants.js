import React from 'react';
import { FormattedMessage } from 'react-intl';

export const LINE_FIELDS_MAP = {
  poLineNumber: 'poLineNumber',
  poLineDescription: 'poLineDescription',
  quantityPhysical: 'cost.quantityPhysical',
  quantityElectronic: 'cost.quantityElectronic',
  currency: 'cost.currency',
  listUnitPrice: 'cost.listUnitPrice',
  listUnitPriceElectronic: 'cost.listUnitPriceElectronic',
  additionalCost: 'cost.additionalCost',
  discount: 'cost.discount',
  poLineEstimatedPrice: 'cost.poLineEstimatedPrice',
  fundDistribution: 'fundDistribution',
  productIds: 'productIdentifier',
  materialTypePhysical: 'materialType',
  materialTypeE: 'materialTypeE',
  acquisitionMethod: 'acquisitionMethod',
  titleOrPackage: 'titleOrPackage',
  publicationDate: 'publicationDate',
  publisher: 'publisher',
  rush: 'rush',
  vendorAccount: 'vendorDetail.vendorAccount',
  // instructions: 'vendorDetail.instructions',
};

export const LINE_FIELDS_LABELS = {
  [LINE_FIELDS_MAP.poLineNumber]: <FormattedMessage id="ui-orders.poLine.number" />,
  [LINE_FIELDS_MAP.poLineDescription]: <FormattedMessage id="ui-orders.poLine.poLineDescription" />,
  [LINE_FIELDS_MAP.quantityPhysical]: <FormattedMessage id="ui-orders.cost.quantityPhysical" />,
  [LINE_FIELDS_MAP.quantityElectronic]: <FormattedMessage id="ui-orders.cost.quantityElectronic" />,
  [LINE_FIELDS_MAP.currency]: <FormattedMessage id="ui-orders.cost.currency" />,
  [LINE_FIELDS_MAP.listUnitPrice]: <FormattedMessage id="ui-orders.cost.listPrice" />,
  [LINE_FIELDS_MAP.listUnitPriceElectronic]: <FormattedMessage id="ui-orders.cost.unitPriceOfElectronic" />,
  [LINE_FIELDS_MAP.additionalCost]: <FormattedMessage id="ui-orders.cost.additionalCost" />,
  [LINE_FIELDS_MAP.discount]: <FormattedMessage id="ui-orders.cost.discount" />,
  [LINE_FIELDS_MAP.poLineEstimatedPrice]: <FormattedMessage id="ui-orders.cost.estimatedPrice" />,
  [LINE_FIELDS_MAP.fundDistribution]: <FormattedMessage id="ui-orders.poLine.fundDistribution" />,
  [LINE_FIELDS_MAP.productIds]: <FormattedMessage id="ui-orders.itemDetails.productId" />,
  [LINE_FIELDS_MAP.materialTypePhysical]: <FormattedMessage id="ui-orders.filter.materialType.physical" />,
  [LINE_FIELDS_MAP.materialTypeE]: <FormattedMessage id="ui-orders.filter.materialType.electronic" />,
  [LINE_FIELDS_MAP.acquisitionMethod]: <FormattedMessage id="ui-orders.poLine.acquisitionMethod" />,
  [LINE_FIELDS_MAP.titleOrPackage]: <FormattedMessage id="ui-orders.orderLineList.titleOrPackage" />,
  [LINE_FIELDS_MAP.publicationDate]: <FormattedMessage id="ui-orders.itemDetails.publicationDate" />,
  [LINE_FIELDS_MAP.publisher]: <FormattedMessage id="ui-orders.itemDetails.publisher" />,
  [LINE_FIELDS_MAP.rush]: <FormattedMessage id="ui-orders.poLine.rush" />,
  [LINE_FIELDS_MAP.vendorAccount]: <FormattedMessage id="ui-orders.search.vendorDetail.vendorAccount" />,
  'vendorDetail.instructions': <FormattedMessage id="ui-orders.vendor.instructions" />,
};
