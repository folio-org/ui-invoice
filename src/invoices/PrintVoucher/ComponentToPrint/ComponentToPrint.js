import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import get from 'lodash/get';

import {
  Col,
  KeyValue,
  Label,
  NoValue,
  Row,
} from '@folio/stripes/components';
import {
  AmountWithCurrencyField,
  FolioFormattedDate,
  KeyValueInline,
} from '@folio/stripes-acq-components';

import { VOUCHER_STATUS_LABEL } from '../../../common/constants';
import {
  LINE_FIELDS_LABELS,
  LINE_FIELDS_MAP,
} from './constants';
import css from './ComponentToPrint.css';

function PrintValue({ path, source, currency }) {
  const value = get(source, path, null);

  switch (path) {
    case LINE_FIELDS_MAP.amount:
      return (
        <AmountWithCurrencyField
          currency={currency}
          amount={value}
        />
      );
    case LINE_FIELDS_MAP.fundDistributions:
      return source[path]?.map(({ code }) => code).join(', ') || null;
    default:
      return value ?? <NoValue />;
  }
}

PrintValue.propTypes = {
  path: PropTypes.string.isRequired,
  source: PropTypes.object.isRequired,
  currency: PropTypes.string.isRequired,
};

const ComponentToPrint = ({ dataSource: { batchGroup, voucher, vendor, invoice, voucherLines } = {} }) => {
  return (
    <>
      <Row>
        <Col xs={6}>
          <Label>
            <FormattedMessage id="ui-invoice.voucher.print.voucher" />
          </Label>
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.settings.batchGroups.labelSingular" />}
            value={batchGroup?.name}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.voucherDate" />}
            value={<FolioFormattedDate value={voucher?.voucherDate} utc={false} />}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.voucherNumber" />}
            value={voucher?.voucherNumber}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.status" />}
            value={VOUCHER_STATUS_LABEL[voucher?.status] || voucher?.status}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.accountingCode" />}
            value={voucher?.accountingCode}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.vendor.name" />}
            value={vendor?.name}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.vendor.code" />}
            value={vendor?.code}
          />
        </Col>
        <Col xs={6}>
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.accountNo" />}
            value={voucher.accountNo}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.vendorInvoiceNo" />}
            value={invoice.vendorInvoiceNo}
          />

          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.information.invoiceDate" />}
            value={<FolioFormattedDate value={invoice?.invoiceDate} />}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.voucher.print.invoiceTotal" />}
            value={<AmountWithCurrencyField amount={invoice?.total} currency={invoice?.currency} />}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.currency" />}
            value={voucher?.invoiceCurrency}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.invoice.details.voucher.exchangeRate" />}
            value={voucher?.exchangeRate}
          />
          <KeyValueInline
            label={<FormattedMessage id="ui-invoice.voucher.print.voucherTotal" />}
            value={<AmountWithCurrencyField amount={voucher?.amount} currency={voucher?.systemCurrency} />}
          />
        </Col>
      </Row>

      {voucherLines?.map((line, i) => {
        return (
          <div key={line.id}>
            <Row>
              <Col xs={12}>
                <KeyValue
                  label={LINE_FIELDS_LABELS[LINE_FIELDS_MAP.lineNumber]}
                >
                  {i + 1}
                </KeyValue>
              </Col>
            </Row>
            <Row className={css.voucherLineBlock}>
              {Object.keys(LINE_FIELDS_MAP).map((col) => {
                if (col === LINE_FIELDS_MAP.lineNumber) return null;

                return (
                  <Col xs={3} key={col}>
                    <KeyValue
                      label={LINE_FIELDS_LABELS[LINE_FIELDS_MAP[col]]}
                    >
                      <PrintValue currency={voucher?.systemCurrency} path={LINE_FIELDS_MAP[col]} source={line} />
                    </KeyValue>
                  </Col>
                );
              })}
            </Row>
          </div>
        );
      })}
    </>
  );
};

ComponentToPrint.propTypes = {
  dataSource: PropTypes.object,
};

export default ComponentToPrint;
