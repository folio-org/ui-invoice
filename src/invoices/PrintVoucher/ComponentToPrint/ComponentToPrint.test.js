import React from 'react';
import { render } from '@testing-library/react';

import '../../../../test/jest/__mock__';

import {
  batchGroup,
  batchVoucher,
  batchVoucherLine,
  invoice,
} from '../../../../test/jest/fixtures';

import ComponentToPrint from './ComponentToPrint';

const defaultProps = {
  dataSource: {
    batchGroup,
    invoice,
    vendor: {
      name: 'Amazon',
      code: 'AMZ',
    },
    voucher: batchVoucher,
    voucherLines: [batchVoucherLine],
  },
};
const renderComponentToPrint = (props = defaultProps) => render(
  <ComponentToPrint {...props} />,
);

describe('PrintVoucher -> ComponentToPrint', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should render correct structure', () => {
    const { asFragment } = renderComponentToPrint();

    expect(asFragment()).toMatchSnapshot();
  });
});
