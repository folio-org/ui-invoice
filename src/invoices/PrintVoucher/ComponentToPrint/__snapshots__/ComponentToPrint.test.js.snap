// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrintVoucher -> ComponentToPrint should render correct structure 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col-xs-6"
    >
      <label
        class="label"
      >
        ui-invoice.voucher.print.voucher
      </label>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.settings.batchGroups.labelSingular
        </span>
        : Amherst College (AC)
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.voucher.voucherDate
        </span>
        : 2019-12-06
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.voucher.voucherNumber
        </span>
        : 1023
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.voucher.status
        </span>
        : ui-invoice.voucher.status.paid
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.voucher.accountingCode
        </span>
        : 0206
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.vendor.name
        </span>
        : Amazon
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.vendor.code
        </span>
        : AMZ
      </div>
    </div>
    <div
      class="col-xs-6"
    >
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.voucher.accountNo
        </span>
        : 
        <span>
          -
        </span>
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.vendorInvoiceNo
        </span>
        : edi
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.information.invoiceDate
        </span>
        : 2021-06-08
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.voucher.print.invoiceTotal
        </span>
        : 
        <span>
          -
        </span>
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.currency
        </span>
        : USD
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.invoice.details.voucher.exchangeRate
        </span>
        : 1
      </div>
      <div>
        <span
          class="label inlineLabel"
        >
          ui-invoice.voucher.print.voucherTotal
        </span>
        : $23.45
      </div>
    </div>
  </div>
  <div>
    <div
      class="row"
    >
      <div
        class="col-xs-12"
      >
        <div
          class="kvRoot"
        >
          <div
            class="kvLabel"
          >
            ui-invoice.voucher.print.lineNumber
          </div>
          <div
            class="kvValue"
            data-test-kv-value="true"
          >
            1
          </div>
        </div>
      </div>
    </div>
    <div
      class="voucherLineBlock row"
    >
      <div
        class="col-xs-3"
      >
        <div
          class="kvRoot"
        >
          <div
            class="kvLabel"
          >
            ui-invoice.voucher.voucherLines.fundCode
          </div>
          <div
            class="kvValue"
            data-test-kv-value="true"
          >
            ,,
          </div>
        </div>
      </div>
      <div
        class="col-xs-3"
      >
        <div
          class="kvRoot"
        >
          <div
            class="kvLabel"
          >
            ui-invoice.voucher.voucherLines.externalAccountNumber
          </div>
          <div
            class="kvValue"
            data-test-kv-value="true"
          >
            54321098
          </div>
        </div>
      </div>
      <div
        class="col-xs-3"
      >
        <div
          class="kvRoot"
        >
          <div
            class="kvLabel"
          >
            ui-invoice.voucher.voucherLines.amount
          </div>
          <div
            class="kvValue"
            data-test-kv-value="true"
          >
            $23.45
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
